Review #3

sergeizaitcev opened the pull request:
Внимание!

~~Это не законченная версия. В данном PR отсутствует основная логика `gophermart` и произведенный рефакторинг `storage` для `gophermart`.~~

UPD: вся логика сервисов `gophermart` и `accrual` реализована.

UPD: добавлены e2e-тесты.

sergeizaitcev committed:
Initial commit

sergeizaitcev committed:
Add `goose`

sergeizaitcev committed:
Add project layout

koteyye committed:
add storage, db, migrations

koteyye committed:
Мелки правки по ревью

sergeizaitcev committed:
Merge pull request #2 from sergeizaitcev/1-add-an-interfaces-of-repository

add storage, db, migrations

koteyye committed:
добавлены CRUD для Auth + тесты

koteyye committed:
Добавлены CRUD для Auth и тесты бд

koteyye committed:
правки по ревью

sergeizaitcev committed:
Merge pull request #7 from sergeizaitcev/5-add-crud-db-for-auth

5 add crud db for auth

sergeizaitcev committed:
Rebasing

sergeizaitcev committed:
Fix `docker-compose.yaml`

sergeizaitcev committed:
Add sign package

sergeizaitcev committed:
Add the auth service

sergeizaitcev committed:
Add todo

sergeizaitcev committed:
Review fix

sergeizaitcev committed:
Fix init tables

sergeizaitcev committed:
Fix `makefile`

sergeizaitcev committed:
Fix `errors.go`

sergeizaitcev committed:
Update the auth storage

sergeizaitcev committed:
Reduce the cost from 12 to 4

koteyye committed:
Merge pull request #9 from sergeizaitcev/4-add-auth

sergeizaitcev committed:
Add luhn package

sergeizaitcev committed:
Update comment

koteyye committed:
Merge pull request #11 from sergeizaitcev/10-luhn

koteyye committed:
add crud orders

koteyye committed:
add crud orders

sergeizaitcev committed:
Add config

sergeizaitcev committed:
Merge pull request #17 from sergeizaitcev/16-add-config

Add config

koteyye committed:
Правки по ревью

koteyye committed:
Merge branch '8-add-crud-orders-db' of github.com:sergeizaitcev/gophermart into 8-add-crud-orders-db

koteyye committed:
Правки по ревью

koteyye committed:
правки по ревью

koteyye committed:
Правки по ревью

koteyye committed:
Мелкие правки по ревью

sergeizaitcev committed:
Merge pull request #15 from sergeizaitcev/8-add-crud-orders-db

add crud orders

sergeizaitcev committed:
Update migrations

sergeizaitcev committed:
Update config

sergeizaitcev committed:
Update storage errors

sergeizaitcev committed:
Update storage

sergeizaitcev committed:
Update models

sergeizaitcev committed:
Add server

sergeizaitcev committed:
Fix migrations

koteyye committed:
Merge pull request #18 from sergeizaitcev/13-auth-handlers

Auth handlers

koteyye committed:
добавлены crud balance и balance operation

koteyye committed:
добавлены миграции accrual

koteyye committed:
конфиг и миграции accrual

koteyye committed:
небольшие правки по миграциям

koteyye committed:
Мелкие правки по интерфейсам

koteyye committed:
Правки по ревью

sergeizaitcev committed:
Update github actions

sergeizaitcev committed:
Update github actions

Commit:
Merge remote-tracking branch 'upstream/main' into develop

sergeizaitcev committed:
Merge pull request #20 from sergeizaitcev/19-add-crud-balance

koteyye committed:
добавил userID для инкремента и декремента

sergeizaitcev committed:
Merge pull request #24 from sergeizaitcev/23-add-userid-increment-and-decrement

добавил userID для инкремента и декремента

koteyye committed:
правки по ревью + обновил мок

koteyye committed:
Merge pull request #22 from sergeizaitcev/21-accrual-config-and-migrations

21 accrual config and migrations

koteyye committed:
Добавлено создание и обновление orders

koteyye committed:
Добавлено создание и обновление orders

sergeizaitcev committed:
Add monetary package

sergeizaitcev committed:
Add a client for accrual

koteyye committed:
Merge pull request #27 from sergeizaitcev/26-add-a-client-for-accrual

26 add a client for accrual

koteyye committed:
добавлены crud

koteyye committed:
добавлены crud

koteyye committed:
добавлены crud

koteyye committed:
добавлены crud

koteyye committed:
добавлен GetOrderWithGoodsByNumber

koteyye committed:
добавлен GetOrderWithGoodsByNumber

koteyye committed:
Merge remote-tracking branch 'origin/25-add-crud-accrual-db' into 25-add-crud-accrual-db

# Conflicts:
#	go.sum

koteyye committed:
добавлен GetOrderWithGoodsByNumber

koteyye committed:
правки по структуре

koteyye committed:
рефакторинг storage.go

Commit:
изменен тип для полей price и accrual

koteyye committed:
добавлен сервис и хендлер для создания заказа

sergeizaitcev committed:
Merge pull request #28 from sergeizaitcev/25-add-crud-accrual-db

25 add crud accrual db

koteyye committed:
добавлен хендлер для регистрации заказа

sergeizaitcev committed:
Merge pull request #30 from sergeizaitcev/29-add-method-create-order

29 add method create order

koteyye committed:
добавлен метод для создания match

koteyye committed:
add get orders

sergeizaitcev committed:
Merge pull request #32 from sergeizaitcev/31-add-handler-get-order-and-create-match

31 add handler get order and create match

gonozov0 commented on internal/accrual/models/errors.go:
```diff
@@ -0,0 +1,15 @@
+
+import "errors"
+
+// Ошибки storage
```
Давайте вместо комментов в эти пакеты по errors.go файлам и раскидаем их

gonozov0 commented on internal/accrual/service/service.go:
```diff
@@ -0,0 +1,214 @@
+type Service struct {
+	storage storage.Storage
+	termCh chan struct{}
+	wg sync.WaitGroup
```
Стоит передавать по указателю, иначе возможны сюрпризы, что где-то создался новый экземпляр и по итогу wait будет работать бесконечно

gonozov0 commented on internal/gophermart/service/accrual/client.go:
Клиент для внешней системы (Accrual Client) и клиент для базы данных (Storage) - с точки зрения архитектуры кода должны находиться на одном уровне, в проектах, где я применяю DDD - это infrastructure слой. Соответственно я бы предложил вам на одном уровне с пакетом service создать пакет infra и туда поместить accuralclient и storage

gonozov0 commented on internal/gophermart/service/accrual/client.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+// Client определяет HTTP-клиент для запросов в accrual.
+type Client struct {
```
Тут нужен интерфейс и соответственно моки, как и для Storage. Для того, чтобы можно тестировать бизнес логику без походов во внешний мир

gonozov0 reviewed (changes requested):
Прекрасное начало 👍
Отписал пару комментариев по улучшению кода

Ожидаю, что вы еще пофиксите автотесты в ci
Также не увидел интеграционных тестов. Под интеграционными тестами я понимаю api тесты, которые тестируют handler-ы и бизнес логику (в вашем случае это service), при этом не ходят во внешний мир, т.е используют MockStorage и MockAccrualClient (в случае gophermart, в случае accrual только MockStorage)

Написание таких тестов - это важная часть при работе на любом реальном рабочем проекте, поэтому важно получить опыт их написания в финальном проекте

Commit:
добавлены unit тесты handler + правки accrual по ревью

koteyye committed:
доделал тесты

sergeizaitcev committed:
Update codebase

sergeizaitcev committed:
big commit

sergeizaitcev committed:
commit

sergeizaitcev committed:
Merge pull request #35 from sergeizaitcev/34-add-unit-tests-accrual

34 add unit tests accrual

sergeizaitcev committed:
Merge pull request #36 from sergeizaitcev/update

Update

sergeizaitcev committed:
Generate mocks

sergeizaitcev committed:
Fix linter

koteyye committed:
fix panic in pipeline

koteyye committed:
fix wg

sergeizaitcev committed:
Fix authorization

sergeizaitcev committed:
Fix accrual client

sergeizaitcev committed:
Fix gophermart initialized

sergeizaitcev committed:
Fix order processing

koteyye committed:
fix migration

koteyye committed:
fix accrual and gophermart db

sergeizaitcev committed:
Remove gophermarttest

koteyye committed:
fix unit test storage

koteyye committed:
добавил правки в handler

sergeizaitcev committed:
Fix storage

sergeizaitcev committed:
Merge remote-tracking branch 'upstream/fix' into fix

sergeizaitcev committed:
Fix autotests

sergeizaitcev committed:
Fix secret key path

sergeizaitcev committed:
Add default secret key

sergeizaitcev committed:
Disable secret_key.txt

sergeizaitcev committed:
Fix actions

sergeizaitcev committed:
Revert config

sergeizaitcev committed:
Fix actions

sergeizaitcev committed:
Logging a config

sergeizaitcev committed:
Revert "Revert config"

This reverts commit b40499949fa2ddcc84cd9319ee2332d1480cf626.

sergeizaitcev committed:
Fix secret path

sergeizaitcev committed:
Done!

sergeizaitcev committed:
Merge pull request #37 from sergeizaitcev/fix

Fix

sergeizaitcev committed:
Update fifo tests

koteyye committed:
refactor migration and add service test

koteyye committed:
refactor migration and add service test

koteyye committed:
refactor migration and service test

koteyye committed:
refactor migration and service test

koteyye committed:
add only digital order accrual

koteyye committed:
add only digital order accrual

sergeizaitcev committed:
Refactoring of the gophermart

sergeizaitcev committed:
Merge remote-tracking branch 'upstream/38-deep-refactoring' into 38-deep-refactoring

koteyye committed:
fix test

koteyye committed:
fix test

sergeizaitcev committed:
Merge pull request #39 from sergeizaitcev/38-deep-refactoring

38 deep refactoring

gonozov0 reviewed (changes requested):
Прекрасная работа, весь код даже покрыли юнитами 👍

Осталась последняя из важных вещей, что я требую в финальном проекте - это написание интеграционных или end-to-end (e2e) тестов.
Под интеграционными тестами я подразумеваю тест на бизнес логику приложения (вызов ручек + проверка результатов), но без походов во внешний мир. То есть для Storage и клиента к внешней системе заказов должны быть определены интерфейсы и реализована их Mock/InMemory имплементация, которая будет использоваться в этих тестах.
Под e2e - тоже тесты на бизнес логику, но в этот раз уже с походами во все внешние сервисы, то есть для их работы нужно будет отдельно поднять локально базу данных и систему заказов, указать верные адреса, порты и т.д.

В рамках этого проекта мне важно увидеть 1 интеграционный или e2e тест на всю бизнес логику заказа, то есть:
1. дергается ручка создания заказа
2. в фоне происходит поход во внешнюю систему и обновляется статус заказа на Processed
3. в фоне происходит начисление баллов на счет пользователя
4. в тесте происходит проверка, что заказ обновлен и увеличен баланс пользователя

sergeizaitcev committed:
Add e2e tests

Add a throttling package

Add a httputil package

Update the accrual client

Add e2e tests

...

sergeizaitcev committed:
Merge "develop" branch into "e2e"

sergeizaitcev committed:
Merge pull request #40 from sergeizaitcev/e2e

E2E

gonozov0 reviewed (approved):
По итогу получилась отличная работа 👍
Хорошо разбили проект по слоям, сгруппировали код, который можно переиспользовать в дальнейших проектах
Есть юнит тесты и самое важное интеграционные тесты, если бы нужно было релизить в прод, то можно было бы смело это делать
В проект будет легко погружаться новым людям и вносить какие-либо изменения!)
